shawk.md

```diff
@@ -0,0 +1,92 @@
+# ShAWK - Extracting Data at the Command Line
+Shesmu can become a repository of data, but it can still be necessary to access
+that data in an exploratory way. Shesmu provides a companion command line tool
+for extracting and filtering data using an AWK-like syntax: `shawk`. The
+tool uses an HTTP endpoint `/extract` so it is also possible to use `curl` or
+`wget` to access the same data.
+
+## Configuring ShAWK
+The command line tool has a configuration file that allows creating reusable
+configuration. The location of the file depends on the OS, so use `--help` to
+see the path. The configuration file is a YAML file. All configuration
+parameters are optional:
+
+```
+---
+aliases:
+  prod: "https://jrhacker@shesmu-prod.example.com/"
+  dev: "https://jrhacker@shesmu-dev.example.com/"
+default_host: prod
+default_input_format: cerberus_fp
+default_output_format: TSV
+prepared_columns:
+  tissue: "tissue_type, tissue_origin, tissue_name, tissue_region, tissue_prep"
+```
+
+The tool needs to know the URL for each Shesmu instance, but this can be
+cumbersome, so `aliases` allows specifying short names for URLs. The full URL
+can be provided to the `-H` or `--host` switch, but an alias can also be used.
+If no host is provided on the command line, the `default_host` is used, which
+can be a URL or alias.
+
+Data will be extracted from a particular input format, and, again it can be
+specified on the command line using the `-i` or `--input` switch. If omitted,
+the `default_input_format` is used for the input format.
+
+Similarly, the data can be made into several formats, described later, and the
+`-f` or `--format` switch determines the output format, but this can be omitted
+and a default format used.
+
+ShAWK is about extracting and manipulating columns. The `prepared_columns`
+allows defining reusable groups of columns that can be inserted into a query.
+More details in the following section.
+
+## Queries
+The data to extract is specified using an AWK-like structure with olive syntax.
+A query is made of one or more rules. Each rule specifies the columns it
+generates. If multiple rules are specified, they must produce the same columns
+in the same order. A simple rule can specify columns to copy:
+
+    shawk -i cerberus_fp -f TSV '{project, library_name, tissue_type}'
+
+Columns can also be _gangs_ using the `@`-like object construction in an olive:
```

Comment:
Sorry, I should have been clearer in my question. I actually want to know what "@-like object" means

Reply:
When you have a gang in Shesmu, you can go `{@some_gang}` to create a tuple with that data or `"{@some_gang}"` to create a string. Same idea here where `@some_gang` will be interpreted as a bunch of columns.

Reply:
It is technically different because we're producing columns, not objects. Like, if `some_gang` had columns `foo` and `bar`, requesting:
```
{@some_gang, other = 3}
```
will produce columns:
`foo, bar, other`

while:
```
{x = {@some_gang}, other = 3}
```
is legal for JSON and XML and will produce something like:
```
<record>
  <x>
    <foo>...</foo>
    <bar>...</bar>
  </x>
  <other>3</other>
</record>
```

So, they are distinct, but if the distinction is too subtle, there might be a way to clarify.

Reply:
ok, so the '@ column construction' then?

Reply:
Fixed.